File: src/StepParser/Core/ProductionMatch.cs
namespace StepParser.Core;

/// <summary>
/// Represents a match result for a production.
/// </summary>
public class ProductionMatch
{
    private readonly Production _production;
    private readonly string _value;
    private readonly int _startPosition;
    private readonly int _endPosition;

    /// <summary>
    /// Initializes a new instance of the <see cref="ProductionMatch"/> class.
    /// </summary>
    /// <param name="production">The production that matched.</param>
    /// <param name="value">The matched value.</param>
    /// <param name="startPosition">The start position of the match.</param>
    /// <param name="endPosition">The end position of the match.</param>
    public ProductionMatch(Production production, string value, int startPosition, int endPosition)
    {
        _production = production;
        _value = value;
        _startPosition = startPosition;
        _endPosition = endPosition;
    }

    /// <summary>
    /// Gets the production that matched.
    /// </summary>
    public Production Production => _production;

    /// <summary>
    /// Gets the matched value.
    /// </summary>
    public string Value => _value;

    /// <summary>
    /// Gets the start position of the match.
    /// </summary>
    public int StartPosition => _startPosition;

    /// <summary>
    /// Gets the end position of the match.
    /// </summary>
    public int EndPosition => _endPosition;

    /// <summary>
    /// Gets the length of the match.
    /// </summary>
    public int Length => _endPosition - _startPosition;

    /// <summary>
    /// Returns a string representation of this production match.
    /// </summary>
    /// <returns>A string representation.</returns>
    public override string ToString()
    {
        return $"ProductionMatch({_production.Name}, '{_value}', {_startPosition}-{_endPosition})";
    }
}

File: old_code/src/utils/ProductionMatch.ts
/**
 * Represents a production match result.
 */
export class ProductionMatch {
  private production: Production;
  private matches: ProductionMatch[];
  private value: string;
  private startPosition: number;
  private endPosition: number;

  /**
   * Creates a new ProductionMatch instance.
   * @param production The production that matched
   * @param value The matched value
   * @param startPosition The start position of the match
   * @param endPosition The end position of the match
   */
  constructor(production: Production, value: string, startPosition: number, endPosition: number) {
    this.production = production;
    this.matches = [];
    this.value = value;
    this.startPosition = startPosition;
    this.endPosition = endPosition;
  }

  /**
   * Gets the production.
   */
  public getProduction(): Production {
    return this.production;
  }

  /**
   * Gets the child matches.
   */
  public getMatches(): ProductionMatch[] {
    return this.matches;
  }

  /**
   * Adds a child match.
   * @param match The match to add
   */
  public addMatch(match: ProductionMatch): void {
    this.matches.push(match);
  }

  /**
   * Gets the matched value.
   */
  public getValue(): string {
    return this.value;
  }

  /**
   * Gets the start position of the match.
   */
  public getStartPosition(): number {
    return this.startPosition;
  }

  /**
   * Gets the end position of the match.
   */
  public getEndPosition(): number {
    return this.endPosition;
  }

  /**
   * Creates a string representation of this production match.
   */
  public toString(): string {
    return `ProductionMatch(${this.production.getName()}, "${this.value}", ${this.startPosition}-${this.endPosition})`;
  }
}

// Import required classes
import { Production } from './Production';
